File: day-23/part-1/thomren.rs
use std::env::args;
use std::time::Instant;

fn main() {
    let now = Instant::now();
    let output = run(&args().nth(1).expect("Please provide an input"));
    let elapsed = now.elapsed();
    println!("_duration:{}", elapsed.as_secs_f64() * 1000.);
    println!("{}", output);
}

fn run(input: &str) -> isize {
    let grid: Vec<&[u8]> = input.lines().map(|l| l.as_bytes()).collect();
    let start = (0, 1);
    let end = (grid.len() - 1, grid[0].len() - 2);

    let mut stack = vec![(start, 0)];
    let mut visited = vec![vec![false; grid[0].len()]; grid.len()];
    let mut max = 0;
    while let Some((pos, dist)) = stack.pop() {
        // println!("{:?} {}", pos, dist);
        if pos == end {
            max = max.max(dist);
            continue;
        }
        if dist == -1 {
            visited[pos.0][pos.1] = false;
            continue;
        }
        if visited[pos.0][pos.1] {
            continue;
        }
        visited[pos.0][pos.1] = true;

        // will backtrack once all the paths from pos have been explored
        stack.push((pos, -1));

        for neighbor in get_neighbors(&grid, pos) {
            if !visited[neighbor.0][neighbor.1] {
                stack.push((neighbor, dist + 1));
            }
        }
    }
    max
}

fn get_neighbors(grid: &Vec<&[u8]>, pos: (usize, usize)) -> Vec<(usize, usize)> {
    let mut neighbors = vec![];
    if pos.0 > 0 && (grid[pos.0 - 1][pos.1] == b'.' || grid[pos.0 - 1][pos.1] == b'^') {
        neighbors.push((pos.0 - 1, pos.1));
    }
    if pos.0 < grid.len() - 1 && (grid[pos.0 + 1][pos.1] == b'.' || grid[pos.0 + 1][pos.1] == b'v')
    {
        neighbors.push((pos.0 + 1, pos.1));
    }
    if pos.1 > 0 && (grid[pos.0][pos.1 - 1] == b'.' || grid[pos.0][pos.1 - 1] == b'<') {
        neighbors.push((pos.0, pos.1 - 1));
    }
    if pos.1 < grid[0].len() - 1
        && (grid[pos.0][pos.1 + 1] == b'.' || grid[pos.0][pos.1 + 1] == b'>')
    {
        neighbors.push((pos.0, pos.1 + 1));
    }
    neighbors
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn run_test() {
        assert_eq!(
            run("#.#####################
#.......#########...###
#######.#########.#.###
###.....#.>.>.###.#.###
###v#####.#v#.###.#.###
###.>...#.#.#.....#...#
###v###.#.#.#########.#
###...#.#.#.......#...#
#####.#.#.#######.#.###
#.....#.#.#.......#...#
#.#####.#.#.#########v#
#.#...#...#...###...>.#
#.#.#v#######v###.###v#
#...#.>.#...>.>.#.###.#
#####v#.#.###v#.#.###.#
#.....#...#...#.#.#...#
#.#########.###.#.#.###
#...###...#...#...#.###
###.###.#.###v#####v###
#...#...#.#.>.>.#.>.###
#.###.###.#.###.#.#v###
#.....###...###...#...#
#####################.#"),
            94
        )
    }
}

File: day-23/part-1/thomren.py
from collections import defaultdict, deque
from tool.runners.python import SubmissionPy


class ThomrenSubmission(SubmissionPy):
    def run(self, s: str):
        """
        :param s: input in string format
        :return: solution flag
        """
        grid = [[c for c in line] for line in s.split("\n")]
        start = (0, next(i for i, c in enumerate(grid[0]) if c == "."))
        end = (len(grid) - 1, next(i for i, c in enumerate(grid[-1]) if c == "."))

        stack = [(start, 0)]
        visited = set()
        max_dist = 0
        while stack:
            pos, dist = stack.pop()
            if pos == end:
                max_dist = max(max_dist, dist)
                continue
            elif dist == -1:
                visited.remove(pos)
                continue
            elif pos in visited:
                continue
            visited.add(pos)

            # will backtrack once all the paths from pos have been explored
            stack.append((pos, -1))

            for neighbor in get_neighbors(grid, pos):
                if neighbor not in visited:
                    stack.append((neighbor, dist + 1))

        return max_dist


def get_neighbors(grid: list[list[str]], pos: tuple[int, int]) -> list[tuple[int, int]]:
    for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
        x, y = (pos[0] + dx, pos[1] + dy)
        if (
            not 0 <= x < len(grid)
            or not 0 <= y < len(grid[0])
            or grid[x][y] == "#"
            or (grid[x][y] == ">" and (dx, dy) != (0, 1))
            or (grid[x][y] == "<" and (dx, dy) != (0, -1))
            or (grid[x][y] == "^" and (dx, dy) != (-1, 0))
            or (grid[x][y] == "v" and (dx, dy) != (1, 0))
        ):
            continue
        yield (x, y)


def test_thomren():
    """
    Run `python -m pytest ./day-23/part-1/thomren.py` to test the submission.
    """
    assert (
        ThomrenSubmission().run(
            """
#.#####################
#.......#########...###
#######.#########.#.###
###.....#.>.>.###.#.###
###v#####.#v#.###.#.###
###.>...#.#.#.....#...#
###v###.#.#.#########.#
###...#.#.#.......#...#
#####.#.#.#######.#.###
#.....#.#.#.......#...#
#.#####.#.#.#########v#
#.#...#...#...###...>.#
#.#.#v#######v###.###v#
#...#.>.#...>.>.#.###.#
#####v#.#.###v#.#.###.#
#.....#...#...#.#.#...#
#.#########.###.#.#.###
#...###...#...#...#.###
###.###.#.###v#####v###
#...#...#.#.>.>.#.>.###
#.###.###.#.###.#.#v###
#.....###...###...#...#
#####################.#
""".strip()
        )
        == 94
    )
